### PR TITLE
105393 validate checklist

### PR DESCRIPTION
**Pull request by janid1967**
Until we have checklist and tag data available in Completion DB, Main DB is master for these data.
So we ask endpoint against Main to get checklist and tag data, to be able to validate tag when performing Punch commands.
We use cache so we don't have to ask Main endpoint more than once pr request

### CODE REVIEW

File: src/Equinor.ProCoSys.Completion.Command/Validators/CheckListValidators/CheckListValidator.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    public CheckListValidator(ICheckListCache checkListCache, IPlantProvider plantProvider)
+    {
+        _checkListCache = checkListCache;
+        _plantProvider = plantProvider;
+    }
+
+    public async Task<bool> ExistsAsync(Guid checkListGuid)
+    {
+        var proCoSys4CheckList = await _checkListCache.GetCheckListAsync(_plantProvider.Plant, checkListGuid);
+        return proCoSys4CheckList != null;
```
Comment (jsoi-eq) [inline src/Equinor.ProCoSys.Completion.Command/Validators/CheckListValidators/CheckListValidator.cs:22]: is not null
For all instances in file

File: src/Equinor.ProCoSys.Completion.Command/Validators/PunchItemValidators/PunchItemValidator.cs [hunk 2]
```diff
@@ -6,29 +6,36 @@
...
+    public async Task<bool> TagOwningPunchItemIsVoidedAsync(Guid punchItemGuid, CancellationToken cancellationToken)
     {
-        // todo #103935 update code below to query tag table to check if tag for punch item is voided .. remember unit test
-        //var tag = await (from pi in _context.QuerySet<PunchItem>()
-        //    join t in _context.QuerySet<Tag>() on pi.TagId equals t.Id 
-        //    where pi.Guid == punchItemGuid
-        //    select t).SingleOrDefaultAsync(cancellationToken);
-        //return tag is not null && tag.IsVoided;
-        return Task.FromResult(false);
+        var punchItem = await GetPunchItem(punchItemGuid, cancellationToken);
```
Comment (jsoi-eq) [inline src/Equinor.ProCoSys.Completion.Command/Validators/PunchItemValidators/PunchItemValidator.cs:31]: GetPunchItemAsync?

File: src/Equinor.ProCoSys.Completion.Command/Validators/PunchItemValidators/PunchItemValidator.cs [hunk 3]
```diff
@@ -6,29 +6,36 @@
...
+
+        if (punchItem == null)
```
Comment (jsoi-eq) [inline src/Equinor.ProCoSys.Completion.Command/Validators/PunchItemValidators/PunchItemValidator.cs:33]: is null

File: src/Equinor.ProCoSys.Completion.ForeignApi/MainApi/CheckList/CheckListCache.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        _cacheManager = cacheManager;
+        _checkListApiService = checkListApiService;
+    }
+
+    public async Task<ProCoSys4CheckList?> GetCheckListAsync(string plant, Guid checkListGuid)
+        => await _cacheManager.GetOrCreate(
+            CheckListGuidCacheKey(checkListGuid),
+            async () =>
+            {
+                var checkList = await _checkListApiService.GetCheckListAsync(plant, checkListGuid);
```
Comment (jsoi-eq) [inline src/Equinor.ProCoSys.Completion.ForeignApi/MainApi/CheckList/CheckListCache.cs:23]: Can be simplified to:
async () => await _checkListApiService.GetCheckListAsync(plant, checkListGuid)

File: src/Equinor.ProCoSys.Completion.WebApi/Authorizations/ContentAccessChecker.cs [hunk 5]
```diff
@@ -31,10 +30,10 @@ public async Task<bool> HasCurrentUserAccessToCheckListAsync(Guid checkListGuid)
         }
 
         var plant = _plantProvider.Plant;
-        var checkList = await _checkListApiService.GetCheckListAsync(plant, checkListGuid);
+        var checkList = await _checkListCache.GetCheckListAsync(plant, checkListGuid);
         if (checkList is null)
         {
-            throw new InValidCheckListException($"CheckList '{checkListGuid}' is not a valid CheckList in '{plant}'");
+            throw new Exception($"CheckList '{checkListGuid}' is not a valid CheckList in '{plant}'");
```
Comment (jsoi-eq) [inline src/Equinor.ProCoSys.Completion.WebApi/Authorizations/ContentAccessChecker.cs:36]: Is it not valid, or does it not exist?
I.e. if the only check is if it exists or not, then error message should indicate that.
If there are other concerns in play, i.e. voided and/or other business rules, then using "valid" is fine.

File: src/Equinor.ProCoSys.Completion.Command/Validators/PunchItemValidators/PunchItemValidator.cs [hunk 6]
```diff
@@ -6,29 +6,36 @@
...
 
 namespace Equinor.ProCoSys.Completion.Command.Validators.PunchItemValidators;
 
 public class PunchItemValidator : IPunchItemValidator
 {
     private readonly IReadOnlyContext _context;
+    private readonly ICheckListValidator _checkListValidator;
 
-    public PunchItemValidator(IReadOnlyContext context) => _context = context;
+    public PunchItemValidator(IReadOnlyContext context, ICheckListValidator checkListValidator)
```
Comment (jsoi-eq) [inline src/Equinor.ProCoSys.Completion.Command/Validators/PunchItemValidators/PunchItemValidator.cs:18]: Should a validator call another validator?
My grasp of the business concepts are a bit hazy, but if a PunchItem contains checklists, and the entry point is always the PunchItem, then I guess it's fine to delegate to another validator for the sub objects.

File: src/tests/Equinor.ProCoSys.Completion.WebApi.IntegrationTests/TestFactory.cs [hunk 7]
```diff
@@ -335,6 +339,8 @@ private void SetupProCoSysServiceMocks()
                 Email = "noreply@pcs.com",
                 ServicePrincipal = true
             }));
+        _checkListApiServiceMock.GetCheckListAsync(PlantWithAccess, CheckListGuid)
+            .Returns(new ProCoSys4CheckList("RC", false, ProjectGuidWithAccess));
```
Comment (jsoi-eq) [inline src/tests/Equinor.ProCoSys.Completion.WebApi.IntegrationTests/TestFactory.cs:343]: Is "RC" a string to be used other places, or is it never referenced anywhere else?
If used other places maybe consider a constant.

**Review by jsoi-eq — COMMENTED** [review]
Awaiting feedback.

File: src/Equinor.ProCoSys.Completion.ForeignApi/MainApi/CheckList/CheckListCache.cs (hunk 4 above)
Comment (janid1967) [reply src/Equinor.ProCoSys.Completion.ForeignApi/MainApi/CheckList/CheckListCache.cs:23]: Easier to read, and set breakpoint as is .... won't fix

File: src/Equinor.ProCoSys.Completion.Command/Validators/PunchItemValidators/PunchItemValidator.cs (hunk 6 above)
Comment (janid1967) [reply src/Equinor.ProCoSys.Completion.Command/Validators/PunchItemValidators/PunchItemValidator.cs:18]: I don't see any problem with code as is .. A PunchItem is ALWAYS owned by a checklist

File: src/tests/Equinor.ProCoSys.Completion.WebApi.IntegrationTests/TestFactory.cs (hunk 7 above)
Comment (janid1967) [reply src/tests/Equinor.ProCoSys.Completion.WebApi.IntegrationTests/TestFactory.cs:343]: Will be fixed in next PR ... I'm planning to make more integration tests around the restriction security concept

File: src/Equinor.ProCoSys.Completion.WebApi/Authorizations/ContentAccessChecker.cs (hunk 5 above)
Comment (janid1967) [reply src/Equinor.ProCoSys.Completion.WebApi/Authorizations/ContentAccessChecker.cs:36]: ...and the code is temporary while procosys 4 is master for chechlist. In future, checklist will also be mastered in Completion. Then the ICheckListCache will be thrown and replaced with a checklistrepo class, getting checklist in local db.

So both "not valid" or "not found" is fine for me .. "Not found" is maybe the best

**Review by jsoi-eq — APPROVED** [review]

**Pull request merged**